source/2018-09-18-the-git-fixup-workflow.html.markdown

```diff
@@ -0,0 +1,213 @@
+
+$ git commit -m "add user component's index view"
+
+# coding the user component's edit view
+
+$ git commit -m "add user component's edit view"
+
+# updating the component's index view so it actually works
+
+$ git commit --fixup <COMMIT HASH OF THE COMMIT "add user component's index view">
+
+# a typo is noticed in the index view
+
+$ git commit --fixup <COMMIT HASH OF THE COMMIT "add user component's index view">
+
+# fix a bug in the view rendering
+
+$ git commit -m "fix view rendering"
+
+$ rubocop
+$ yarn lint
+
+# fix linting errors, assuming the linting error was introduced in the commit
+# "add user component's edit view"
+
+$ git commit --fixup <COMMIT HASH OF THE COMMIT "add user component's edit view">
+
+# team members do a code review on the PR / MR
+
+$ git commit -m "fix import in index"
+git commit --fixup <COMMIT HASH OF THE COMMIT "add user component's index view">
+```
+
+The resulting git log looks like this:
+
+```
+fixup! add user component's index view
+fixup! add user component's edit view
+fix view rendering
+fixup! add user component's index view
+fixup! add user component's index view
+add user component's edit view
+add user component's index view
+```
+
+Instead of regularly committing (and also writing a commit message) we use the
+`--fixup` flag to create fixup commits.
+
+Now, before merging the merge request we issue an interactive rebase with the `--autosquash` flag.
+
+> **Note:**
+> GitLab marks MRs that include fixup commits as `WIP`, so merging is
+> prevented if you haven't squashed.
+
+```bash
+$ git rebase --interactive --autosquash <COMMIT HASH OF THE COMMIT "add user component's index view">~1
+```
+
+We'll enter interactive rebase mode with the relevant commits already marked for fixup:
+
+```
+pick 364003b add user component's index view
+fixup afa6970 fixup! add user component's index view
+fixup 1da5d7c fixup! add user component's index view
+fixup c998b08 fixup! add user component's index view
+pick d9731b5 add user component's edit view
+fixup 33d6080 fixup! add user component's edit view
+pick 73e8a6a fix view rendering
+```
+
+The resulting git history looks like this:
+
+```
+fix view rendering
+add user component's edit view
+add user component's index view
+```
+
+The end result is a very consistent and readable git history, where every other
+developer knows what happens, without the uninteresting sidesteps you took
+while developing.
+
+
+## Summary
+
+In summary the following two essential git commands are necessary to make use
+of the git fixup workflow:
+
+```bash
+$ git commit --fixup <COMMIT THAT NEEDS FIXING>
+$ git rebase -i --autosquash <FIRST COMMIT THAT NEEDS FIXING>~1
+```
+
+
+## Bonus: Create git aliases
+
+If you start adopting this workflow you'll have to type a lot in you terminal.
+At this point it makes sense to create some git aliases to ease your brain and
+hands.
+
+```conf
+[alias]
+    cif = commit --fixup
+    ri = rebase -i --autosquash
+```
```

Comment:
More bonus:

- `git config --global rebase.autoSquash true` (still requires `-i` though)
- https://github.com/tombruijn/dotfiles/blob/master/bin/git-fixup, displays a list of recent commits so you can just enter a number for the commit, instead of having to copy the hash

Reply:
Nice! The git-fixup script doesn't have a license, so I think we shouldn't reference it.

Reply:
@toupeira WDYT reg. the license?

Reply:
Oh well ;-)

Reply:
Not sure, linking couldn't hurt I guess? But I'm also ok with not mentioning it at all.

It seems code without a license [can be considered public domain if it's very short](https://github.com/github/choosealicense.com/issues/196), but IANAL :)

Reply:
I'll drop this, I'm not comfortable with no-license stuff.